Refactor Docker Stack

burnoutprojects opened the pull request:
I update the development and production docker stack with security and minimalism in mind.

I'm open for advice and comments.

P.S.: Happy new year, happy 2024 🎉

burnoutprojects committed:
Remove old docker implementation

burnoutprojects committed:
Add dockerignore file

burnoutprojects committed:
Replace ts-node with tsx package

burnoutprojects committed:
Create multi-stage image file and entrypoint file

* Dockerfile: Set of instructions for creating development and production image
* entrypoint.sh: This file override the default entrypoint of the image. Install npm dependencies and run npm script based of enviroment

burnoutprojects committed:
Update entrypoint.sh

* Add message for app running mode (ex. development, production)
* Add build script to production mode

burnoutprojects committed:
.gitignore: Ignore .dev-data folder

burnoutprojects committed:
tsconfig.json: Include src folder, exclude node_modules folder

burnoutprojects committed:
Remove nodemon.json config file

burnoutprojects committed:
Remove nodemon package

burnoutprojects committed:
package.json: Update docker scripts

burnoutprojects committed:
.dockerignore: Cleanup entries

burnoutprojects committed:
Update npm scripts and entrypoint scripts file

burnoutprojects committed:
Dockerfile: Update multi-stage image

* Add Builder step
* Add Development step
* Add Production step

burnoutprojects committed:
Add new docker compose files for development and production use

* dev.compose.yaml - For development only use
* prod.compose.yaml - For production only use

burnoutprojects committed:
Move @swc/helpers and dotenv to dependencies

In current structure of docker if not present, the app is not starting because missing dependencies

burnoutprojects committed:
entrypoint.sh: Update script to run with sh from bash

burnoutprojects committed:
Dockerfile: Update the images structure

burnoutprojects committed:
prod.compose.yaml: Comment out the database port for extra security

casperiv0 commented on package.json:
```diff
@@ -4,19 +4,25 @@
-    "dev": "npx prisma db push && nodemon",
-    "build": "npx prisma generate && npx swc ./src -d dist",
-    "start": "npx prisma migrate deploy && node dist/index.js",
+    "dev": "prisma migrate dev && tsx watch src/index.ts",
```
Why change this? Please keep `nodemon` here.

burnoutprojects replied:
I think nodemon is an extra dependency we don't need. We can use `--watch` flag from node v18+ and is more efficient from nomemon. This is my opinion of cource and I want to talk about it.

casperiv0 commented on tsconfig.json:
```diff
@@ -22,24 +29,18 @@
     /* Advanced Options */
     "skipLibCheck": true /* Skip type checking of declaration files. */,
     "forceConsistentCasingInFileNames": true /* Disallow inconsistently-cased references to the same file. */
-  },
-  "ts-node": {
-    "swc": true
```
Please revert these changes. SWC speeds up a lot of things, so let's keep it

casperiv0 commented on prod.compose.yaml:
```diff
@@ -0,0 +1,30 @@
+    command: -p ${DB_PORT:-55545}
+    env_file:
+      - .env
+    # ports:
+    #   - "${DB_PORT:-55545}:${DB_PORT:-55545}"
```
```suggestion
    ports:
      - "${DB_PORT:-55545}:${DB_PORT:-55545}"
```

burnoutprojects replied:
If we make ports visible to outside world, is a big security issue if the user is not config firewall correctly.

casperiv0 commented on prod.compose.yaml:
```diff
@@ -0,0 +1,30 @@
+    depends_on:
+      - postgres
+
+volumes:
+  snailybot-db-data:
```
This volume will be removed when the Docker container is removed, this can lead to unwanted things such as dataloss. Please use a local volume like before.

burnoutprojects replied:
If we keep the local mount folder, the build is failed on Linux hosts, with this error

`failed to solve: rpc error: code = Unknown desc = error from sender: open <project_path>/snailycad-bot/dev-data: permission denied`

casperiv0 commented on dev.compose.yaml:
```diff
@@ -0,0 +1,32 @@
+    depends_on:
+      - postgres
+
+volumes:
+  snailybot-dev-db-data:
```
This volume will be removed when the Docker container is removed, this can lead to unwanted things such as dataloss. Please use a local volume like before.

burnoutprojects replied:
If we keep the local mount folder, the build is failed on Linux hosts, with this error

`failed to solve: rpc error: code = Unknown desc = error from sender: open <project_path>/snailycad-bot/dev-data: permission denied`

burnoutprojects committed:
Update .gitignore and .dockerignore

- Add .data and .dev-data in .gitignore file
- Add .data and .dev-data in .dockerignore file

burnoutprojects committed:
Dockerfile: Remove useless label

burnoutprojects committed:
package.json: Update dev script to use ts-node

- Re-add ts-node in dev dependencies
- Update the dev script to use ts-node with node --watch flag

burnoutprojects committed:
tsconfig.json: Re-add ts-node config

burnoutprojects committed:
entrypoint.js: Update docker entrypoint script

burnoutprojects committed:
dev.compose.yaml: Update dev compose file

- Add .dev-data folder as volume
- Remove docker volume

burnoutprojects committed:
dev.compose.yaml: Update dev compose file

- Add .dev-data folder as volume
- Remove docker volume
- Uncomment ports

burnoutprojects committed:
package.json: Update scripts

- Add docker:clean script
- Update docker-dev:start script
- Update docker-prod:start script
- Update docker-dev:build script
- Update docker-prod:build script

burnoutprojects committed:
Dockerfile: Change production image

burnoutprojects committed:
Add --silent flag to all yarn install

burnoutprojects committed:
Add init service inside container
